### PR TITLE
fix: hide 5xx error message from client

### PR DESCRIPTION
**Pull request by liangliang4ward**
### Description

hidding 5xx error message from client

Fixes https://github.com/apache/apisix/issues/6699

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/authz-keycloak.lua [hunk 1]
```diff
@@ -572,7 +572,7 @@ local function evaluate_permissions(conf, ctx, token)
     -- Ensure discovered data.
     local err = authz_keycloak_ensure_discovered_data(conf)
     if err then
-        return 500, err
```
Comment (tokers) [inline apisix/plugins/authz-keycloak.lua:575]: Ditto.

Comment (liangliang4ward) [reply apisix/plugins/authz-keycloak.lua:575]: done

File: apisix/plugins/authz-casbin.lua [hunk 2]
```diff
@@ -110,9 +110,9 @@ end
 
 function _M.rewrite(conf, ctx)
     -- creates an enforcer when request sent for the first time
-    local ok, err = new_enforcer_if_need(conf)
+    local ok = new_enforcer_if_need(conf)
```
[the remark this answers is not in the document]
Comment (liangliang4ward) [reply apisix/plugins/authz-casbin.lua]: done

File: t/plugin/authz-keycloak.t [hunk 3]
```diff
@@ -337,8 +339,7 @@ passed
 GET /t
 --- response_body
 false
---- error_log
```
Comment (moonming) [inline t/plugin/authz-keycloak.t:340]: I think we also need to check the error log?

File: apisix/plugins/authz-keycloak.lua [hunk 4]
```diff
@@ -581,17 +582,17 @@ local function evaluate_permissions(conf, ctx, token)
...
         -- Resolve URI to resource(s).
         permission, err = authz_keycloak_resolve_resource(conf, ctx.var.request_uri,
                                                           sa_access_token)
 
         -- Check result.
-        if permission == nil then
+        if permission == nil or err then
             -- No result back from resource registration endpoint.
-            return 500, err
+            return 503
```
Comment (shuaijinchao) [inline apisix/plugins/authz-keycloak.lua:595]: ditto

**Comment by leslie-tsang** [comment]
Hello there, I will take over this PR as it lacks of feedback. :)

**Comment by liangliang4ward** [comment]
> Hello there, I will take over this PR as it lacks of feedback. :)

I'm sorry about that, I'm already wait for reply.

now I see you commit , I learn something.

**Comment by leslie-tsang** [comment]
> I'm sorry about that, I'm already wait for reply.
> 
> now I see you commit , I learn something.

Ohhh, I see, Glad to have you back, do you like to continue ?

**Comment by liangliang4ward** [comment]
> > I'm sorry about that, I'm already wait for reply.
> > now I see you commit , I learn something.
> 
> Ohhh, I see, Glad to have you back, do you like to continue ?

let me do this continue;
I thought that recording ERROR logs might result in two error logs for the test case. Because it's already been recorded once.

Now that I see your commit, I think logging WARN would be a better way.

**Comment by leslie-tsang** [comment]
> let me do this continue; I thought that recording ERROR logs might result in two error logs for the test case. Because it's already been recorded once.
> 
> Now that I see your commit, I think logging WARN would be a better way.

Ok, please check those review comments carefully and fix them according to the comments. :)

Would you take some time to finish it today ?

**Comment by liangliang4ward** [comment]
> > let me do this continue; I thought that recording ERROR logs might result in two error logs for the test case. Because it's already been recorded once.
> > Now that I see your commit, I think logging WARN would be a better way.
> 
> Ok, please check those review comments carefully and fix them according to the comments. :)
> 
> Would you take some time to finish it today ?

ok

**Comment by leslie-tsang** [comment]
Hello there, @liangliang4ward, There are still some comment need to be take care of. I will reopen them for you.

File: t/plugin/authz-keycloak.t (hunk 3 above)
Comment (leslie-tsang) [reply t/plugin/authz-keycloak.t:340]: @liangliang4ward Please check this out.

**Comment by liangliang4ward** [comment]
@leslie-tsang hi, ci run failed. but not relate this PR code;

**Comment by leslie-tsang** [comment]
> @leslie-tsang hi, ci run failed. but not relate this PR code;

I will check it out. Thanks for you contribution.

**Comment by leslie-tsang** [comment]
> > @leslie-tsang hi, ci run failed. but not relate this PR code;
> 
> I will check it out. Thanks for you contribution.

CI pass now. 

> I think we also need to check the error log?

Please add the test case to cover the change.

**Comment by liangliang4ward** [comment]
hi, I will close this pr, I find the master already modify some code relate with "authz-keycloak.lua" ,some logic is changed.

**Comment by tzssangglass** [comment]
hi @liangliang4ward , if what you think `already modify some code relate with "authz-keycloak.lua"` is this PR: #6854, I think there is some misunderstanding.

 #6854 does not conflict with this PR, so you can go on and reopen this PR.

**Comment by liangliang4ward** [comment]
some code I don't know how process;
like function "authz_keycloak_get_token_endpoint"
some code call this  return : 
```lua
return 503, {message = err}  
````
and other code return:
```lua
 return 500, "Unable to determine token endpoint."
```
and refer from https://github.com/apache/apisix/issues/6699
will be: return 503;

**Comment by tzssangglass** [comment]
> some code I don't know how process; like function "authz_keycloak_get_token_endpoint" the master new code call this function return :
> 
> ```lua
> return 503, {message = err}  
> ```
> 
> and other code return:
> 
> ```lua
>  return 500, "Unable to determine token endpoint."
> ```
> 
> and refer from #6699 will be: return 503;

we can just use 503? cc @spacewander

**Comment by spacewander** [comment]
It's recommended to return 503 as people keep thinking the 500 status code is the error of APISIX.

**Pull request closed without merging**